torch/csrc/autograd/python_function.cpp

```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   Py_INCREF((PyObject*)self);
   return std::shared_ptr<PyFunction>(&self->cdata, Decref());
 }
+
+namespace torch { namespace autograd {
+
+using memo_map = std::unordered_map<Node*, variable_list>;
```

Comment:
So we're going to remember all the intermediate outputs? That's not a very good idea, because we save _**a lot**_ of memory, by only keeping what we want. `Variable`s can be huge.